File: java_samples/ResultAbbreviation.java
package ai.cochlear.example;

import ai.cochl.sense.model.SenseEvent;
import ai.cochl.sense.model.WindowHop;

import java.util.*;

class ResultAbbreviation {
    static class BufferValue {
        private final double intervalMargin;
        private final double startTime;
        private final double endTime;

        BufferValue(double im, double startTime, double endTime) {
            this.intervalMargin = im;
            this.startTime = startTime;
            this.endTime = endTime;
        }

        public double getIntervalMargin() {
            return intervalMargin;
        }

        public double getStartTime() {
            return startTime;
        }

        public double getEndTime() {
            return endTime;
        }
    }

    static String _kTagNameOther = "Others";

    private final boolean enabled;
    private final float hopSize;
    private final int defaultIntervalMargin;
    private final LinkedHashMap<String, Integer> tagsIM;

    private final LinkedHashMap<String, BufferValue> buffer;
    private double minimumAcceptableMargin;

    private final StringBuilder stringBuilder1 = new StringBuilder();
    private final StringBuilder stringBuilder2 = new StringBuilder();

    public ResultAbbreviation(boolean enabled, int defaultIntervalMargin, WindowHop windowHop, LinkedHashMap<String, Integer> tagsIM) {
        this.enabled = enabled;
        this.defaultIntervalMargin = defaultIntervalMargin;

        if (windowHop == WindowHop._0_5S) {
            this.hopSize = 0.5f;
        } else if (windowHop == WindowHop._1S) {
            this.hopSize = 1.0f;
        } else {
            throw new IllegalArgumentException("Hop size can only be 0.5 or 1");
        }

        this.tagsIM = tagsIM;
        this.buffer = new LinkedHashMap<>();

        this.minimumAcceptableMargin = 0;
        if (this.defaultIntervalMargin == 0 && this.hopSize == 0.5d) {
            this.minimumAcceptableMargin = -0.5d;
        }
    }

    String minimizeDetails(List<SenseEvent> results, boolean endOfFile) {
        if (!enabled) {
            return "Result Abbreviation is currently disabled";
        }

        stringBuilder1.setLength(0);

        for (SenseEvent result : results) {
            String line = minimizeDetailsFrame(result);
            if (line.isEmpty()) {
                continue;
            }

            stringBuilder1.append(line);
        }

        if (endOfFile) {
            for (Map.Entry<String, BufferValue> entry : buffer.entrySet()) {
                String tagName = entry.getKey();
                double fromTime = entry.getValue().getStartTime();
                double toTime = entry.getValue().getEndTime();

                stringBuilder1.append(getLine(fromTime, toTime, tagName));
            }
        }

        return stringBuilder1.toString();
    }

    String minimizeDetailsFrame(SenseEvent senseEvent) {
        if (!enabled) {
            return "Result Abbreviation is currently disabled";
        }

        stringBuilder2.setLength(0);

        double startTime = senseEvent.getStartTime();
        double endTime = senseEvent.getEndTime();
        ArrayList<String> treatedTags = new ArrayList<>();

        for (int i = 0; i < senseEvent.getTags().size(); i++) {
            String tagName = senseEvent.getTags().get(i).getName();

            if (Objects.equals(tagName, _kTagNameOther)) {
                continue;
            }

            double fromTime;
            if (buffer.containsKey(tagName)) {
                fromTime = buffer.get(tagName).getStartTime();
                buffer.put(tagName, new BufferValue(getIntervalMarginByTag(tagName), fromTime, endTime));
            } else {
                buffer.put(tagName, new BufferValue(getIntervalMarginByTag(tagName), startTime, endTime));
            }

            treatedTags.add(tagName);
        }

        ArrayList<String> tagsToRemove = new ArrayList<>();
        for (Map.Entry<String, BufferValue> entry : buffer.entrySet()) {
            String tagName = entry.getKey();
            if (treatedTags.contains(tagName)) {
                continue;
            }

            double fromTime = entry.getValue().getStartTime();
            double toTime = entry.getValue().getEndTime();

            double im = entry.getValue().getIntervalMargin();
            im -= hopSize;

            if (im < minimumAcceptableMargin) {
                stringBuilder2.append(getLine(fromTime, toTime, tagName));
                tagsToRemove.add(tagName);
            } else {
                buffer.put(tagName, new BufferValue(im, fromTime, toTime));
            }
        }

        for (String s : tagsToRemove) {
            buffer.remove(s);
        }

        return stringBuilder2.toString();
    }

    int getIntervalMarginByTag(String tagName) {
        if (tagsIM.containsKey(tagName)) {
            return tagsIM.get(tagName);
        }

        return defaultIntervalMargin;
    }

    String getFormattedTime(double time) {
        String formattedTime = Double.toString(time);
        return formattedTime.substring(0, formattedTime.indexOf('.') + 2);
    }

    String getLine(double fromTime, double toTime, String tagName) {
        String from = getFormattedTime(fromTime);
        String to = getFormattedTime(toTime);

        return "\nAt " + from + "-" + to + "s, [" + tagName + "] was detected";
    }
}

File: python_sample/result_abbreviation.py
HALF_SECOND = 0.5  # default
ONE_SECOND = 1


class ResultAbbreviation:
    def __init__(
        self,
        enabled=True,
        default_im=0,
        hop_size=HALF_SECOND,
        tags_im: dict[str, int] = None,
    ):
        self.enabled = enabled
        if hop_size != HALF_SECOND and hop_size != ONE_SECOND:
            raise ValueError("Hop size can only be 0.5 or 1")
        self.hop_size = hop_size
        self.default_im = default_im
        self.tags_im = tags_im
        self._buffer = {}
        self._file_mode = False
        self._tag_name_other = "Others"

    def minimize_details(self, results=[], file_ended=False):
        """
        Used for file inference to loop over the results and provide a single result summary
        """
        if not self.enabled:
            return "Result Abbreviation is currently disabled"

        self._file_mode = True
        output = ""

        for frame_result in results:
            line = self.minimize_details_frame(frame_result)
            if not line:
                continue
            output = self._append_line(output, line)

        if file_ended is True:
            to_time = results[-1]["end_time"]
            for tag, (_, from_time, to_time) in self._buffer.items():
                line = f"At {from_time}-{to_time}s, {tag} was detected"
                output = self._append_line(output, line)

        self._file_mode = False
        return output

    def minimize_details_frame(self, frame_result):
        """
        This can be used as you loop over the results
        It will output a merge of the results in ranges
        """
        if not self.enabled:
            return "Result Abbreviation is currently disabled"

        output = ""
        nb_line = 0

        start_time = frame_result["start_time"]
        end_time = frame_result["end_time"]
        treated_tags = []

        for tag in frame_result["tags"]:
            name = tag["name"]
            if name == self._tag_name_other:
                continue

            if name in self._buffer:
                (_, from_time, _) = self._buffer[name]
                self._buffer[name] = (self._im(name), from_time, end_time)
            else:
                self._buffer[name] = (self._im(name), start_time, end_time)
            treated_tags.append(name)

        tags_to_remove = []
        for tag, (im, from_time, to_time) in self._buffer.items():
            if tag in treated_tags:
                continue

            im -= self.hop_size
            if im < 0:
                line = f"At {from_time}-{to_time}s, {tag} was detected"
                nb_line += 1
                output = self._append_line(output, line)
                tags_to_remove.append(tag)
            else:
                self._buffer[tag] = (im, from_time, to_time)

        for tag in tags_to_remove:
            del self._buffer[tag]

        return output

    def clear_buffer(self):
        self._buffer.clear

    def _im(self, tag):
        if self.tags_im is None:
            return self.default_im
        return self.tags_im.get(tag, self.default_im)

    def _append_line(self, result, line):
        if not result:
            return line
        return f"{result}\n{line}"
